client/rpc-servers/src/lib.rs

```diff
@@ -121,34 +128,42 @@ pub fn start_ws<M: Send + Sync + 'static>(
 	addrs: &[SocketAddr],
 	max_connections: Option<usize>,
 	cors: Option<&Vec<String>>,
-	maybe_max_payload_mb: Option<usize>,
-	module: RpcModule<M>,
+	max_payload_mb: Option<usize>,
+	prometheus_registry: Option<&Registry>,
+	rpc_api: RpcModule<M>,
 	rt: tokio::runtime::Handle,
 ) -> Result<WsServerHandle, anyhow::Error> {
-	let max_request_body_size = maybe_max_payload_mb
+	let max_request_body_size = max_payload_mb
 		.map(|mb| mb.saturating_mul(MEGABYTE))
 		.unwrap_or(RPC_MAX_PAYLOAD_DEFAULT);
 	let max_connections = max_connections.unwrap_or(WS_MAX_CONNECTIONS);
 
-	let mut builder = WsServerBuilder::default()
+	let mut builder = WsServerBuilder::new()
 		.max_request_body_size(max_request_body_size as u32)
 		.max_connections(max_connections as u64)
 		.custom_tokio_runtime(rt.clone());
 
-	log::info!("Starting JSON-RPC WS server: addrs={:?}, allowed origins={:?}", addrs, cors);
-
 	if let Some(cors) = cors {
 		// Whitelist listening address.
 		// NOTE: set_allowed_hosts will whitelist both ports but only one will used.
 		builder = builder.set_allowed_hosts(format_allowed_hosts(addrs))?;
 		builder = builder.set_allowed_origins(cors)?;
 	}
 
-	let server = tokio::task::block_in_place(|| rt.block_on(builder.build(addrs)))?;
-
-	let rpc_api = build_rpc_api(module);
-	let handle = server.start(rpc_api)?;
+	let rpc_api = build_rpc_api(rpc_api);
+
+	let handle = if let Some(prometheus_registry) = prometheus_registry {
+		let metrics = RpcMetrics::new(&prometheus_registry)?;
+		let middleware = RpcMiddleware::new(metrics, "ws".into());
+		let builder = builder.set_middleware(middleware);
+		let server = tokio::task::block_in_place(|| rt.block_on(builder.build(addrs)))?;
```

Comment:
maybe return `builder` from the if/else to avoid repeating `tokio::task::block_in_place && server.start(rpc_api)`

Reply:
I wish. It doesn't work because once you set the middleware they're different types.

Reply:
ah alright